File: data/src/main/java/com/github/anrimian/musicplayer/data/database/entities/folder/FolderEntity.java
package com.github.anrimian.musicplayer.data.database.entities.folder;

import androidx.room.Entity;
import androidx.room.ForeignKey;
import androidx.room.Index;
import androidx.room.PrimaryKey;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;

@Entity(tableName = "folders",
        foreignKeys = {
                @ForeignKey(entity = FolderEntity.class,
                        parentColumns = "id",
                        childColumns = "parentId",
                        onDelete = ForeignKey.CASCADE)
        },
        indices = {
                @Index("parentId")
        }
)
public class FolderEntity {

    @PrimaryKey(autoGenerate = true)
    private long id;

    @Nullable
    private Long parentId;

    @Nonnull
    private String name;

    public FolderEntity(@Nullable Long parentId, @Nonnull String name) {
        this.parentId = parentId;
        this.name = name;
    }

    public void setId(long id) {
        this.id = id;
    }

    public long getId() {
        return id;
    }

    @Nullable
    public Long getParentId() {
        return parentId;
    }

    @Nonnull
    public String getName() {
        return name;
    }
}

File: data/src/main/java/com/github/anrimian/musicplayer/data/database/entities/folder/FolderEntity.kt
package com.github.anrimian.musicplayer.data.database.entities.folder

import androidx.room.Entity
import androidx.room.ForeignKey
import androidx.room.Index
import androidx.room.PrimaryKey

@Entity(
    tableName = "folders",
    foreignKeys = [
        ForeignKey(
            entity = FolderEntity::class,
            parentColumns = [ "id" ],
            childColumns = [ "parentId" ],
            onDelete = ForeignKey.CASCADE
        ),
        ForeignKey(
            entity = VolumeEntity::class,
            parentColumns = [ "id" ],
            childColumns = [ "volumeId" ],
            onDelete = ForeignKey.CASCADE
        )
    ],
    indices = [ Index("parentId"), Index("volumeId") ]
)
class FolderEntity(
    @PrimaryKey(autoGenerate = true)
    val id: Long,
    val parentId: Long?,
    val volumeId: Long?, // only non null when parentId is null(root folders)
    val name: String
)
